File: RayTracer/Material/Material.cs
using System;
using System.Collections.Generic;
using System.Drawing;
using System.Linq;
using System.Text;

namespace RayTracer.Material
{
    [Serializable]
    public class Mat
    {

        public Mat()
        {
            Specular = new MyColor();
            Diffuse = new MyColor();
            Emission = new MyColor();
            Shininess = 0f;
        }

        public MyColor Specular
        { get; set; }

        public MyColor Diffuse
        { get; set; }

        public MyColor Emission
        { get; set; }


        float shininess = 0;
        public float Shininess
        {
            get
            { return shininess; }
            set
            { shininess = (value < 0) ? 0 : ((value > 128) ? 128 : value); }
        }

        float refractValue = 0;
        public float RefractValue
        {
            get
            { return refractValue; }
            set
            { refractValue = (value < 0) ? 0 : ((value > 1) ? 1 : value); }
        }

        float refractIndex = 1;
        public float RefractIndex
        {
            get
            { return refractIndex; }
            set
            { refractIndex = (value < 0) ? 0 : value; }
        }

        public Mat Clone()
        {
            return (Mat)this.MemberwiseClone();
        }
    }
}

File: RayTracer - BVH - CUDA/RayTracerCUDA/Material.cpp
#include "Material.h"



Material::Material()
{
}

 void Material::setShininess(float val)
{
	shininess = val;
	if (shininess < 0)
		shininess = 0;
	if (shininess > 128)
		shininess = 128;
}

 void Material::setrefIndex(float val)
{
	refractIndex = val;
	if (refractIndex < 0)
		refractIndex = 0;
	if (refractIndex > 1)
		refractIndex = 1;
}

 void Material::setRefValue(float val)
{
	refractValue = val;
	if (refractValue < 0)
		refractValue = 0;
	if (refractValue > 1)
		refractValue = 1;
}


Material::~Material()
{
}
